nixos-container: fix allow alternative nixos paths

ericbmerritt opened the pull request:
###### Motivation for this change
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This should be completely backwards compatible. It allows the '-f' part
of the nix-env command to be configured. This greatly eases using
nixos-container as part of development where several nixpkgs
repositories might be tested at the same time.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nixos/nixpkgs/15967)

<!-- Reviewable:end -->

ericbmerritt committed:
nixos-container: fix allow alternative nixos paths

This should be completely backwards compatible. It allows the '-f' part
of the nix-env command to be configured. This greatly eases using
nixos-container as part of development where several nixpkgs
repositories might be tested at the same time.

joachifm commented:
cc @edolstra

lethalman commented:
@edolstra looks fine, perhaps it should also be supported for declarative containers? I guess it's not necessary, right? Or do you see potential problems?

ericbmerritt commented:
@lethalman You are probably correct. I use it exclusively for on the fly containers, but I could see good reasons for having it in the declarative containers as well.

lethalman commented:
@ericbmerritt I'm ok with the change, I just wonder if it may break declarative containers at first sight...

ericbmerritt commented:
In the case where no path is passed, it should have the exact same behaviour as before.

ericbmerritt committed:
nixos-container: feature add 'config-file' option

This adds a config file option to nixos-container.pl that makes it quite
a bit easier to use.

ericbmerritt commented:
@lethalman I just added an additional change that makes containers quite a bit easier to use for testing configurations. Both of these should be backwards compatible.

fpletz commented on nixos/modules/virtualisation/nixos-container.pl:
```diff
@@ -68,6 +73,15 @@ sub showHelp {
+
+    if ($extraConfig) {
+        $localExtraConfig = $extraConfig
+    } elsif ($configFile) {
```
I see a potential problem here. If both `--config` and `--config-file` are provided, only the `--config` parameter will be respected and `--config-file` will be ignored silently, right? We should probably either merge `$localExtraConfig` or exit with an error.

fpletz commented:
Apart from my comment, :+1: from me.

ericbmerritt commented:
@fpietz updated

ericbmerritt commented:
is there something more I need to do here?

danbst commented on nixos/modules/virtualisation/nixos-container.pl:
```diff
@@ -34,20 +35,28 @@ sub showHelp {
 
+if (defined $configFile and defined $extraConfig) {
+    die "--config and --config-file are mutually incompatible. " .
+        "Please define on or the other, but not both";
```
typo "on"

joachifm commented:
@ericbmerritt looks like the consensus is yay on this, so if you rebase, I'll merge it.

ericbmerritt commented:
@joachifm  rebased.

edolstra commented:
I don't understand the need for this flag. Why not just set $NIX_PATH to point to the alternative NixOS tree?

ericbmerritt commented:
@edolstra nix_path is ignored by the underlying command.

cleverca22 commented:
nix-env does obey the search path when it encounters a <search> but by default, it loads directly from the channel and never tries to search for things
